src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs

```diff
@@ -2685,5 +2652,524 @@ public static int M()
+  IL_000b:  ret
+}
+""");
+        }
+
+        [Theory]
+        [InlineData("string")]
+        [InlineData("object")]
+        [InlineData("decimal")]
+        [InlineData("nint")]
+        [InlineData("nuint")]
+        [InlineData("System.DateTime")]
+        [InlineData("Test")]
+        public void EmptyArrayCtor_OtherTypes(string type)
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan($$"""
+using System;
+
+class Test
+{
+    public static void Main()
+    {       
+        // inplace inits
+        var s1 = new ReadOnlySpan<{{type}}>(new {{type}}[] { });
+        var s2 = new ReadOnlySpan<{{type}}>(new {{type}}[] { });
+
+        Console.Write(s1.Length == s2.Length);
+
+        // make an instance
+        Console.Write(s1.Length == new ReadOnlySpan<{{type}}>(new {{type}}[] { }).Length);
+    }
+}
+""", TestOptions.ReleaseExe);
+
+            CompileAndVerify(comp, expectedOutput: "TrueTrue", verify: Verification.Passes).VerifyIL("Test.Main", $$"""
+{
+  // Code size       69 (0x45)
+  .maxstack  2
+  .locals init (System.ReadOnlySpan<{{type}}> V_0, //s1
+                System.ReadOnlySpan<{{type}}> V_1, //s2
+                System.ReadOnlySpan<{{type}}> V_2)
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  initobj    "System.ReadOnlySpan<{{type}}>"
+  IL_0008:  ldloca.s   V_1
+  IL_000a:  initobj    "System.ReadOnlySpan<{{type}}>"
+  IL_0010:  ldloca.s   V_0
+  IL_0012:  call       "int System.ReadOnlySpan<{{type}}>.Length.get"
+  IL_0017:  ldloca.s   V_1
+  IL_0019:  call       "int System.ReadOnlySpan<{{type}}>.Length.get"
+  IL_001e:  ceq
+  IL_0020:  call       "void System.Console.Write(bool)"
+  IL_0025:  ldloca.s   V_0
+  IL_0027:  call       "int System.ReadOnlySpan<{{type}}>.Length.get"
+  IL_002c:  ldloca.s   V_2
+  IL_002e:  initobj    "System.ReadOnlySpan<{{type}}>"
+  IL_0034:  ldloc.2
+  IL_0035:  stloc.2
+  IL_0036:  ldloca.s   V_2
+  IL_0038:  call       "int System.ReadOnlySpan<{{type}}>.Length.get"
+  IL_003d:  ceq
+  IL_003f:  call       "void System.Console.Write(bool)"
+  IL_0044:  ret
+}
+""");
+        }
+
+        [Theory]
+        [InlineData("string")]
+        [InlineData("object")]
+        [InlineData("decimal")]
+        [InlineData("nint")]
+        [InlineData("nuint")]
+        [InlineData("System.DateTime")]
+        public void UnusedSpan_NothingEmitted(string type)
+        {
+            string csharp = $$"""
+public class Test
+{
+    public static int M()
+    {
+        System.ReadOnlySpan<{{type}}> s = new {{type}}[] { default };
+        return 42;
+    }
+}
+""";
+            var compilation = CreateCompilationWithMscorlibAndSpan(csharp, TestOptions.ReleaseDll);
+            var verifier = CompileAndVerify(compilation, verify: Verification.Passes);
+            verifier.VerifyIL("Test.M", """
+{
+  // Code size        3 (0x3)
+  .maxstack  1
+  IL_0000:  ldc.i4.s   42
+  IL_0002:  ret
+}
+""");
+        }
     }
```

Comment:
>} [](http://example.com/codeflow?start=4&length=1)

We probably should have a test that enumerates all possible constant types and makes sure compilation and runtime behavior isn't broken.